### ISSUE TITLE
no-restricted-imports exact pattern match is not possible

### ISSUE THREAD

**Issue by yvele** [eslint/eslint#13857]
**Tell us about your environment**

* **ESLint Version:** v7.13.0
* **Node Version:**  v12.19.0
* **npm Version:** v6.14.8

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

@babel/eslint-parser

**What are you trying to achieve?**

A little background: I'm trying to forbid importing full lodash library `import _ from lodash` and instead force to import explicit lodash sub modules like so `import foo from "lodash/foo"`.

**What did you expect to happen?**

I'm expecting this rule

```js
rules : {
  "no-restricted-imports" : ["error", {
    patterns : [
        // Forbid complete lodash import
        "lodash",
        // Allow lodash sub module import
        "!lodash/*",
    ]
}
```

With this code

```js
import lodash from "lodash";
import cloneDeep from "lodash/cloneDeep";
```

To ONLY throw:

> 1:1 error  'lodash' import is restricted from being used by a pattern  no-restricted-imports

But it's the problem is that line 2 is also reported as an error:

> 1:2  error  'lodash/cloneDeep' import is restricted from being used by a pattern  no-restricted-imports

I also tried  `"!lodash/**"`, `"!lodash/cloneDeep"` as workaround but neither works..

**Are you willing to submit a pull request to fix this bug?**

Why not

**Comment by mdjermanovic** [comment]
Hi @yvele, thanks for the issue!

`patterns` option works per the [gitignore specification](https://git-scm.com/docs/gitignore), so this is intended behavior (emphasis mine):

> An optional prefix "!" which negates the pattern; any matching file excluded by a previous pattern will become included again. __It is not possible to re-include a file if a parent directory of that file is excluded__.

You can use other configuration options to disallow only the exact match:

`"no-restricted-imports": ["error", "lodash"]` - [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIFwibG9kYXNoXCJdKi9cblxuaW1wb3J0IGxvZGFzaCBmcm9tIFwibG9kYXNoXCI7IC8vIGVycm9yXG5pbXBvcnQgY2xvbmVEZWVwIGZyb20gXCJsb2Rhc2gvY2xvbmVEZWVwXCI7IC8vIG9rIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

`"no-restricted-imports": ["error", { name: "lodash" }]` - [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHsgbmFtZTogXCJsb2Rhc2hcIiB9XSovXG5cbmltcG9ydCBsb2Rhc2ggZnJvbSBcImxvZGFzaFwiOyAvLyBlcnJvclxuaW1wb3J0IGNsb25lRGVlcCBmcm9tIFwibG9kYXNoL2Nsb25lRGVlcFwiOyAvLyBvayIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

`"no-restricted-imports": ["error", { paths: ["lodash"] }]` - [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHsgcGF0aHM6IFtcImxvZGFzaFwiXSB9XSovXG5cbmltcG9ydCBsb2Rhc2ggZnJvbSBcImxvZGFzaFwiOyAvLyBlcnJvclxuaW1wb3J0IGNsb25lRGVlcCBmcm9tIFwibG9kYXNoL2Nsb25lRGVlcFwiOyAvLyBvayIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

`"no-restricted-imports": ["error", { paths: [{ name: "lodash" }] }]` - [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHsgcGF0aHM6IFt7IG5hbWU6IFwibG9kYXNoXCIgfV0gfV0qL1xuXG5pbXBvcnQgbG9kYXNoIGZyb20gXCJsb2Rhc2hcIjsgLy8gZXJyb3JcbmltcG9ydCBjbG9uZURlZXAgZnJvbSBcImxvZGFzaC9jbG9uZURlZXBcIjsgLy8gb2siLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by yvele** [comment]
Wow thank you so much for your time @mdjermanovic

I hope this issue will help other people.. I'm closing this.

Thanks for the online demo tip
